### PR TITLE
bpo-43123: revise logic for cr-lf characters in email header names and values

### PR DESCRIPTION
**Pull request by rrhodes**
What?
=====
1. Raise `ValueError` if an email header name includes linefeed or carriage return characters, and
2. allow linefeed and carriage return characters in email header values, so long as a non-space character follows. Where this applies, append a space character to signal continuation of the same header value.

Why?
====
1. To prevent arbitrary MIME headers being injected into email, and
2. to allow a continuation of header values where linefeed or carriage return characters exist, like [this sample message](https://docs.microsoft.com/en-us/previous-versions/office/developer/exchange-server-2010/aa563375(v=exchg.140)).

To quote Martin Ortner, author of this issue:

> I discussed the [header injection] vulnerability in private with one of the email module maintainers and he considers that it's not a vulnerability.

https://bugs.python.org/issue43123

### CODE REVIEW

**Comment by rrhodes** [comment]
@bitdancer, @terryjreedy, @maxking, sorry to ping you out of the blue, but it looks like you're the only three to contribute to the email policy code in the past. Any of you happy to review these changes for me, please?

**Comment by tintinweb** [comment]
sorry for the delay @rrhodes. I'm the one that originally reported the issue (not a python core-dev nor rfc2045 pro 🤓).

IMHO instead of bailing when finding a `\n([!\s]+)` in the value it should be encoded as a value continuation (`\nSOMETEXT` -> `\n<space>SOMETEXT`). I think otherwise this might break content like [this](https://docs.microsoft.com/en-us/previous-versions/office/developer/exchange-server-2010/aa563375(v=exchg.140)):

image

File: Misc/NEWS.d/next/Library/2021-02-07-13-20-40.bpo-43123.cmT1fW.rst [hunk 1]
```diff
@@ -0,0 +1 @@
+Raise ValueError if an email header name includes linefeed or carriage return characters.
```
Comment (tintinweb) [inline Misc/NEWS.d/next/Library/2021-02-07-13-20-40.bpo-43123.cmT1fW.rst]: should mention how values are validated (or encoded)

File: Lib/test/test_email/test_message.py [hunk 2]
```diff
@@ -500,6 +500,29 @@ def set_content(self, msg, *args, **kw):
...
+
+    def test_throws_on_add_invalid_header_name(self):
+        m = self._str_msg('')
+        header_name = 'CUSTOM\nHEADER'
+        header_value = 'data'
+
+        with self.assertRaises(ValueError):
+            m.add_header(header_name, header_value)
+
+    def test_throws_on_add_invalid_header_value(self):
```
Comment (tintinweb) [inline Lib/test/test_email/test_message.py]: maybe add testcase for value continuation (see [content-type/disposition](https://docs.microsoft.com/en-us/previous-versions/office/developer/exchange-server-2010/aa563375(v=exchg.140)))

**Comment by rrhodes** [comment]
Thanks for the feedback @tintinweb, I'll take a look into this in the next couple days. 👍🏻

**Comment by rrhodes** [comment]
@tintinweb I _think_ I've understood the changes you were requesting here: you were happy with the logic for header names, but wanted support for CR and LF characters in header values, which were previously raising a `ValueError`.

Further feedback welcome. If I've misunderstood what is required, I'm happy to discuss this in more detail.

**Comment by ambv** [comment]
I don't understand @tintinweb's suggestion to relax the check for header values. That check is in place *right now* in the `email` library code. What this PR was meant to be about was also introducing a similar check for header *names*. Introducing the latter while relaxing the former looks wrong to me.

**Pull request closed without merging**